api/src/main/java/io/minio/MinioProperties.java

```diff
@@ -31,46 +32,57 @@
 
   private static final Logger LOGGER = Logger.getLogger(MinioProperties.class.getName());
 
+  // These attributes are checked from the manifests in classpath.
+  private static final String META_INF_ATTRIB_IMPLEMENTATION_TITLE = "Implementation-Title";
+  private static final String META_INF_ATTRIB_IMPLEMENTATION_VERSION = "Implementation-Version";
+
+  // this is set from gradle
+  private static final String META_INF_ATTRIB_IMPLEMENTATION_TITLE_VALUE = "minio";
+
   private final AtomicReference<String> version = new AtomicReference<>(null);
 
   public String getVersion() {
     String result = version.get();
-    if (result == null) {
-      synchronized (INSTANCE) {
-        if (version.get() == null) {
-          try {
-            ClassLoader classLoader = getClass().getClassLoader();
-            setMinioClientJavaVersion(classLoader);
-            setDevelopmentVersion();
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "IOException occured", e);
-            version.set("unknown");
-          }
-          result = version.get();
-        }
-      }
+    if (result != null) {
+      return result;
     }
-    return result;
+    setVersion();
+    return version.get();
   }
 
-  private void setDevelopmentVersion() {
-    if (version.get() == null) {
+  private synchronized void setVersion() {
+    ClassLoader classLoader = getClass().getClassLoader();
+    if (classLoader == null) {
       version.set("dev");
+      return;
     }
-  }
 
-  private void setMinioClientJavaVersion(ClassLoader classLoader) throws IOException {
-    if (classLoader != null) {
+    try {
       Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
       while (resources.hasMoreElements()) {
-        Manifest manifest = new Manifest(resources.nextElement().openStream());
-        for (Object k : manifest.getMainAttributes().keySet()) {
-          String versionString = "Implementation-Version";
-          if (k.toString().equals(versionString)) {
-            version.set(manifest.getMainAttributes().getValue((Attributes.Name) k));
+        try (InputStream is = resources.nextElement().openStream()) {
+          Manifest manifest = new Manifest(is);
+          boolean minioManifestFound;
+          for (Map.Entry<Object, Object> entry : manifest.getMainAttributes().entrySet()) {
+            String key = entry.getKey().toString();
+            String value = entry.getValue().toString();
+            minioManifestFound =
```

Comment:
This is not going to work. You are resetting the value on every iteration.

Reply:
Actually we need only below in the `for`
```java
          if (entry.getKey().toString().equals("Implementation-Title") && entry.getValue().toString().equals("minio")) {
            version.set(manifest.getMainAttributes().getValue("Implementation-Version"));
            return;
          }
```

Reply:
Got rid of the for, we can directly pull in the values from the attributes and check